src/main/java/org/wildfly/security/credential/source/OAuth2CredentialSource.java

```diff
@@ -0,0 +1,277 @@
+     * Creates a new {@link Builder} instance in order to configure and build a {@link OAuth2CredentialSource}.
+     *
+     * @param tokenEndpointUrl the token endpoint that will be used to obtain OAuth2 access tokens
+     * @return a new builder instance
+     */
+    public static Builder builder(URL tokenEndpointUrl) {
+        return new Builder(tokenEndpointUrl);
+    }
+
+    private final String grantType;
+    private final URL tokenEndpointUri;
+    private final Consumer<Map<String, String>> consumer;
+    private String scopes;
+    private final Supplier<SSLContext> sslContextSupplier;
+    private final Supplier<HostnameVerifier> hostnameVerifierSupplier;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param grantType the OAuth2 grant type as defined by the specification
+     * @param tokenEndpointUri the OAuth2 Token Endpoint {@link URI}
+     * @param scopes a string with the scope of the access request
+     */
+    private OAuth2CredentialSource(String grantType, URL tokenEndpointUri, String scopes, Supplier<SSLContext> sslContextSupplier, Supplier<HostnameVerifier> hostnameVerifierSupplier) {
+        this(grantType, tokenEndpointUri, (Consumer<Map<String, String>>) stringStringMap -> {}, scopes, sslContextSupplier, hostnameVerifierSupplier);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param grantType the OAuth2 grant type as defined by the specification
+     * @param tokenEndpointUrl the OAuth2 Token Endpoint {@link URL}
+     * @param consumer a callback that can be used to push addition parameters to requests sent to the authorization server
+     * @param scopes a string with the scope of the access request
+     */
+    private OAuth2CredentialSource(String grantType, URL tokenEndpointUrl, Consumer<Map<String, String>> consumer, String scopes, Supplier<SSLContext> sslContextSupplier, Supplier<HostnameVerifier> hostnameVerifierSupplier) {
+        this.grantType = checkNotNullParam("grantType", grantType);
+        this.tokenEndpointUri = checkNotNullParam("tokenEndpointUri", tokenEndpointUrl);
+
+        if (isHttps(tokenEndpointUrl)) {
+            checkNotNullParam("sslContextSupplier", sslContextSupplier);
+        }
+
+        this.consumer = consumer;
+        this.scopes = scopes;
+        this.sslContextSupplier = sslContextSupplier;
+        this.hostnameVerifierSupplier = hostnameVerifierSupplier;
+    }
+
+    @Override
+    public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+        return getCredential(credentialType, algorithmName, parameterSpec) != null ? SupportLevel.SUPPORTED : SupportLevel.UNSUPPORTED;
+    }
+
+    @Override
+    public <C extends Credential> C getCredential(Class<C> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+        if (BearerTokenCredential.class.equals(credentialType)) {
```

Comment:
This should be: ``credentialType.isAssignableFrom(BearerTokenCredential.class)``.  For example I could request a credential of type ``Credential.class`` and I should get an answer if I do.